pkg/target/http.go

```diff
@@ -215,63 +337,150 @@ func AdaptHTTPTargetFunc(f func(c *HTTPTargetConfig) (*HTTPTarget, error)) HTTPT
 
-	safeMessages, oversized := models.FilterOversizedMessages(
+	chunks, oversized := models.GetChunkedMessages(
 		messages,
-		ht.MaximumAllowedMessageSizeBytes(),
+		ht.requestMaxMessages,
+		ht.messageByteLimit-ht.approxTmplSize,
+		ht.requestByteLimit-ht.approxTmplSize,
 	)
 
-	var invalid []*models.Message
-	var failed []*models.Message
-	var sent []*models.Message
+	sent := []*models.Message{}
+	failed := []*models.Message{}
+	invalid := []*models.Message{}
 	var errResult error
+	var hitSetupError bool
 
-	for _, msg := range safeMessages {
-		request, err := http.NewRequest("POST", ht.httpURL, bytes.NewBuffer(msg.Data))
-		if err != nil {
-			errResult = multierror.Append(errResult, errors.Wrap(err, "Error creating request"))
-			failed = append(failed, msg)
-			continue
-		}
-		request.Header.Add("Content-Type", ht.contentType)                // Add content type
-		addHeadersToRequest(request, ht.headers, ht.retrieveHeaders(msg)) // Add headers if there are any
-		if ht.basicAuthUsername != "" && ht.basicAuthPassword != "" {     // Add basic auth if set
-			request.SetBasicAuth(ht.basicAuthUsername, ht.basicAuthPassword)
-		}
-		requestStarted := time.Now().UTC()
-		resp, err := ht.client.Do(request) // Make request
-		requestFinished := time.Now().UTC()
+	for _, chunk := range chunks {
+		grouped := ht.groupByDynamicHeaders(chunk)
 
-		msg.TimeRequestStarted = requestStarted
-		msg.TimeRequestFinished = requestFinished
+		for _, group := range grouped {
+			var reqBody []byte
+			var goodMsgs []*models.Message
+			var badMsgs []*models.Message
+			var err error
 
-		if err != nil {
-			errResult = multierror.Append(errResult, err)
-			failed = append(failed, msg)
-			continue
-		}
-		defer resp.Body.Close()
-		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-			sent = append(sent, msg)
-			if msg.AckFunc != nil { // Ack successful messages
-				msg.AckFunc()
+			if ht.requestTemplate != nil {
+				reqBody, goodMsgs, badMsgs = ht.renderBatchUsingTemplate(group)
+			} else {
+				reqBody, goodMsgs, badMsgs = ht.provideRequestBody(group)
 			}
-		} else {
-			errResult = multierror.Append(errResult, errors.New("Got response status: "+resp.Status))
-			failed = append(failed, msg)
-			continue
+
+			invalid = append(invalid, badMsgs...)
+
+			if len(goodMsgs) == 0 {
+				continue
+			}
+
+			request, err := http.NewRequest("POST", ht.httpURL, bytes.NewBuffer(reqBody))
+
+			if err != nil {
+				panic(err)
+			}
+
+			request.Header.Add("Content-Type", ht.contentType)                        // Add content type
+			addHeadersToRequest(request, ht.headers, ht.retrieveHeaders(goodMsgs[0])) // Add headers if there are any - because they're grouped by header, we just need to pick the header from one message
+			if ht.basicAuthUsername != "" && ht.basicAuthPassword != "" {             // Add basic auth if set
+				request.SetBasicAuth(ht.basicAuthUsername, ht.basicAuthPassword)
+			}
+			requestStarted := time.Now()
```

Comment:
These should explicitly be UTC. I'll push a commit now to fix and check for any others.